### PR TITLE
New storage API

### LINKED ISSUE #139: Refactor storage
Remove all direct store messages and replace with generic methods `storeObject[T]` and `readObject[T]`.

We should factor out the store API from `PetPlugin` and `MailPlugin`.

Also think about more typed storage API.

### PR DESCRIPTION
**Pull request by ForNeVeR**
I've created the more safe storage API. It uses the new flexible and safe paradigm: DAOs (now they're called "repositories") may implement any number of any methods they want (previously they have to implement predefined generic method set: `store`, `read`, `delete`). Callers should pass lambdas to the storage and it will execute these lambdas against newly created repository objects.

To bind repositories to `DBSession`s I've implemented all of them as case classes accepting `DBSession` as parameter. It also helps to better isolate query contexts and to eliminate any possible DAO pool poisoning.

This pull request also eliminates many compiler warnings on generic type casting.

Closes #139.

### CODE REVIEW

**Comment by ForNeVeR** [comment]
cc @rexim

**Comment by ForNeVeR** [comment]
And cc @ttldtor also :3

File: src/main/scala/ru/org/codingteam/horta/core/Core.scala [hunk 1]
```diff
@@ -228,8 +228,9 @@ object Core {
   private def getCommandsDescription(pluginDefinitions: List[(ActorRef, PluginDefinition)]) =
     pluginDefinitions.map(t => t._2.name -> t._2.commands.map(cd => cd.name -> cd.level)).toMap
 
-  private def getStorages(pluginDefinitions: List[(ActorRef, PluginDefinition)]): Map[String, DAO] = {
-    pluginDefinitions.map(_._2).filter(_.dao.isDefined).map(definition => (definition.name, definition.dao.get)).toMap
+  private def getStorages(pluginDefinitions: List[(ActorRef, PluginDefinition)]): Map[String, RepositoryFactory] = {
+    pluginDefinitions.map(_._2).filter(_.repositoryFactory.isDefined).map(
```
Comment (folex) [inline src/main/scala/ru/org/codingteam/horta/core/Core.scala]: Several calls of `.map` in a row is discouraged and may lead to unexpected behaviour.

Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/core/Core.scala]: What kind of unexpected behavior may happen?

File: src/main/scala/ru/org/codingteam/horta/plugins/karma/KarmaPlugin.scala [hunk 2]
```diff
@@ -69,21 +67,18 @@ class KarmaPlugin extends BasePlugin with CommandProcessor {
     Protocol.sendResponse(credential.location, credential, message)
 
   private def showTopKarma(credential: Credential, room:String): Unit = {
-    ((store ? ReadObject(name, GetTopKarma(room))) map {
-      case Some(karma:List[String]) =>
-        "\n" + karma.map(msg => msg).mkString("\n")
-    }).onSuccess({case msg => sendResponse(credential,msg)})
+    withDatabase(_.getTopKarma(room)) map { karma =>
+      val msg = "\n" + karma.map(msg => msg).mkString("\n")
```
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/plugins/karma/KarmaPlugin.scala]: `msg => msg`?!

**Comment by rexim** [comment]
Nothing critical was noticed. I approve that shit! \o/

File: src/main/scala/ru/org/codingteam/horta/plugins/karma/KarmaPlugin.scala (hunk 2 above)
Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/plugins/karma/KarmaPlugin.scala]: That's more: that's `msg = ... (msg => msg)`. Cool one, yeah.

**Comment by ForNeVeR** [comment]
Thanks for the review, guys, we're moving that shit to production!

**Pull request merged**